### PR TITLE
[Ascend] fix: improve Ascend cat/concat dispatch and implementation

### LINKED ISSUE #6500: [Ascend][P1] flag_gems cat/concat 多张量路径触发 Triton-Ascend 编译器段错误（VL 并发 chunked-prefill 场景必崩）
## 现象

昇腾 910C，`flag_gems` rc2 分支 HEAD `4af1e434`（flagtree 0.6.0 + triton 3.5 线），
Qwen3.6-27B **多模态并发**（`qwen3_6_27b_concurrent.py --mode vl`，TP=2，16 并发）时
scheduler 进程 **SIGSEGV 段错误**（exit -11）：

```
Fatal Python error: Segmentation fault
Current thread (most recent call first):
  File ".../triton/backends/ascend/compiler.py", line 124 in ttir_to_linalg   ← 编译器本体段错误
  File ".../triton/backends/ascend/compiler.py", line 830 in <lambda>
  File ".../triton/backends/ascend/spec/triton/compiler/compiler.py", line 285 in compile
  File ".../triton/experimental/tle/__init__.py", line 35 in tle_compile
  File ".../triton/backends/ascend/spec/triton/runtime/jit.py", line 623 in run
  File "/root/FlagGems/src/flag_gems/ops/cat.py", line 181 in _cat_run_kernel   ← flag_gems cat kernel
  File "/root/FlagGems/src/flag_gems/ops/cat.py", line 344 in cat
  File "/root/FlagGems/src/flag_gems/ops/concat.py", line 41 in concat
  File ".../sglang/srt/managers/mm_utils.py", line 653 in _get_chunked_prefill_embedding
  File ".../sglang/srt/managers/mm_utils.py", line 728 in get_embedding_and_mask
  File ".../sglang/srt/managers/mm_utils.py", line 822 in embed_mm_inputs
  File ".../sglang/srt/managers/mm_utils.py", line 1052 in general_mm_embed_routine
  File ".../sglang/srt/models/qwen3_vl.py", line 1272 in forward
  ...
[2026-09-20 15:52:54] Subprocess scheduler_0 (pid=3954860) crashed with exit code -11. Triggering SIGQUIT for cleanup...
```

## 关键结论：与 shape 无关，**多张量 cat 在本环境 100% 编译段错误**

我用真机日志抓到的真实输入 shape，在脱离 sglang 的最小脚本里扫了一遍
（`torch.concat(parts, dim=0)`，bf16，hidden=5120，flag_gems.enable()）：

| 用例 | shapes（各张量行数） | 结果 |
| --- | --- | --- |
| 真机 VL 并发原始调用（15 张量） | 300,100,64,64 ×3 + 300,100,64 | ❌ **SIGSEGV** |
| 单批 4 张量 | 300,100,64,64 | ❌ **SIGSEGV** |
| 单批 3 张量 | 300,100,64 | ❌ **SIGSEGV** |
| 单批 2 张量 | 300,100 | ❌ **SIGSEGV** |
| **单张量** | 300 | ✅ PASS（走 single 分支，不编译 kernel） |
| 两批各 4 张量 | 300,100,64,64 ×2 | ❌ **SIGSEGV** |
| 4 张量**同 shape** | 256,256,256,256 | ❌ **SIGSEGV** |
| 4 张量大方阵 | 1024,1024,1024,1024 | ❌ **SIGSEGV** |

→ **只要走到 `cat_copy_func_kernel`（多张量路径）就崩，跟 shape、跟张量个数的奇偶都无关**；
单张量因为 `_cat_build_working_list` 返回 `mode="single"` 直接返回、不编译 kernel，所以侥幸不崩。

也就是说：**flag_gems 的 `cat`/`concat` 多张量路径在昇腾本环境从未真正可用**，
只是之前没有覆盖到这条路径 —— 单请求 VL 冒烟走的是另一条分支（不触发 chunked-prefill 的 embedding 拼接）。

## 环境

- FlagGems：rc2 分支 HEAD `4af1e434`（含 #6455 等全部已合入修复）
- flagtree：`0.6.0+ascend3.2`（triton 3.5 线）
- sglang：0.5.11 / sgl_kernel_npu 2026.5.1 / plugin 0.2.0-rc2.post1
- 镜像：tree_0.6.0_ascend3.2 老版本镜像（CANN 8.5.0）
- 黑名单：`ascend.yaml` 22 项（含 7 项广播算子，绕 #6238）
- 硬件：昇腾 910C，2 卡 TP=2

### 1) 全链路（约 2 分钟，最快）

```bash
# 昇腾 910C，flagtree 0.6.0 + triton 3.5 + flag_gems rc2 HEAD 4af1e434
export ASCEND_RT_VISIBLE_DEVICES=12,13
export MODEL_PATH=/models/Qwen3.6-27B
export TP_SIZE=2
python3 -u examples/qwen3_6_27b_concurrent.py --mode vl     # → SIGSEGV / RC=137
```

⚠ 只有 **`--mode vl`（并发）** 会崩；单请求 VL 冒烟（`qwen3_6_27b_offline_inference.py`）不触发。
⚠ 只有 **多模态** 会崩；纯文本并发（`--mode text`）不走这条路径，不受影响。

### 2) 脱离 sglang 的最小复现（单次约 18s，不需要加载模型）

> 已自验：逐字跑本段 **3/3 稳定 SIGSEGV**；注释掉 `flag_gems.enable()` 后 **3/3 PASS**（对照）。
> 单次约 17~18s，主要花在 `import torch_npu` + device 初始化上，崩溃发生在编译阶段。

```python
import torch
import torch_npu
import flag_gems

flag_gems.enable()          # 关键：import flag_gems 不注册 op，必须显式 enable
torch_npu.npu.set_device(0) # ⚠ 0 是「可见设备」序号，运行前设 ASCEND_RT_VISIBLE_DEVICES=<空闲卡>

parts = [torch.randn(n, 5120, dtype=torch.bfloat16, device="npu")
         for n in (256, 256, 256, 256)]     # 同 shape 也崩
torch.concat(parts, dim=0)  # → Segmentation fault (exit -11)
```

---

## 影响面

凡是走 **多模态 + chunked prefill** 的 sglang 模型（`mm_utils._get_chunked_prefill_embedding`
结尾的 `torch.concat(embedding_list, dim=0)`），在本环境必然 SIGSEGV。
与 #6249 / PR #6455（linear）**无关**，是独立问题。

## 绕法（已真机验证）

在昇腾 `sglang_fl/dispatch/config/ascend.yaml` 的 `flagos_blacklist` 里加：

```yaml
  - cat      # torch.cat    → cat.py
  - concat   # torch.concat → concat.py:41
```

两个注册名都要加（`flag_gems/__init__.py` 里 `("cat", cat)` 和 `("concat", concat)` 分别注册，
sglang `mm_utils` 走的是 `torch.concat`）。

加上之后，`--mode vl` 从 **SIGSEGV** 变为 **`vl validation passed.`**，
四图全对（red_square→'Red' / cat→'cat' / stop_sign→'red' / digit_seven→'7'），`All concurrent tests passed.`。

## 备注

- 段错误发生在 **Triton-Ascend 编译器 `ttir_to_linalg` 降级阶段**，`.ttir` 已成功生成、
  但 `.json`/`.npubin` 没有产物 —— 可在 `$TRITON_CACHE_DIR` 下看到只剩一个 `cat_copy_func_kernel_4.ttir`。
  同一个 kernel（`BLOCK_X=1024`）在 `--mode text` 全程没被触发过。
- 怀疑是 TTIR → Linalg 降级对 `cat_copy_func_kernel_4` 里那套
  `pid_y == 0/1/2/else` 嵌套 `scf.if` + `idx // (dim_size_in * dim_prod_post)` 结构处理有问题
  （未进一步缩小到具体 op，需要编译器侧同学看）。

### PR DESCRIPTION
**Pull request by modao1234**
### Summary
  
  - 简化 concat.py，直接调用 torch.cat 而非自定义实现
  - 使用 Triton 内核重写 Ascend cat.py 以提升性能
  - 添加完善的输入验证和错误处理
  - 根据拼接维度支持扁平和跨步两种复制模式
  
  ### PR Category
  Operator
  
  ### Type of Change
  Bug Fix | Performance Optimization
  
  ### Description
  
  本 PR 修复了 Ascend cat/concat 算子实现中的调度和性能问题。主要改动包括：
  
  1. **简化 concat.py**：移除自定义实现，直接委托给 `torch.cat`，使调度逻辑更清晰
  2. **基于 Triton 的 cat 实现**：使用 Triton 内核重写 Ascend 后端 cat 算子，包含两个优化路径：
     - 扁平复制内核：用于维度 0 的拼接 
     - 跨步复制内核：用于其他维度的拼接
  3. **改进验证**：添加全面的输入验证，提供清晰的错误信息
  4. **更好的边界情况处理**：正确处理空张量、单张量输入和维度归一化
  
  ### Issue
  
  - Resolves #6500
  
  ### Progress
  
  - [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
  - [x] Change is responded to an issue.
  - [ ] Change is fully covered by a UT.

### CODE REVIEW

**Comment by modao1234** [comment]
不修改通用算子，新增ascend concat特化算子
